### PR TITLE
Comments: redux state management refactor

### PR DESCRIPTION
**Pull request by rodrigoi**
This PR aims to provide a replacement for `state/comments` as the first step of removing `Immutable.JS` from comments state and components.

Changes on the comment block components, post and the reader will follow. These changes will also enable new UI interactions for comments moderation tools, and provides selectors that will enable connecting the comment block components to the redux state.

To test:

```sh
npm run test-client client/state/discussions/test
```
```sh
npm run test-client client/state/selectors/
```
```sh
npm run test-client client/state/data-layer/wpcom/discussions
```

Size is `XL`, which is mostly tests

### CODE REVIEW

**Comment by dmsnell** [comment]
hi! here are a few high-level remarks I would like to share before this is finished:

 - this significant of a change in the fetching (also because it's new code) should be built in the data layer. that should end up reducing the overall size of the PR both in less code needed to wire up the API and in writing tests (probably more files though)
 - love the use of `keyedReducer()` - we can probably double-up there with the comments too.
 - is the "discussions" terminology new here or were we already using it. is there a reason we're not just calling it "comments"?

would be happy to share about the data layer, even do a video on it. in the mean time, you can check out p4TIVU-6qh or our [docs](https://github.com/automattic/wp-calypso/tree/master/client/state/data-layer) on it [in state](https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer/wpcom-http)

**Comment by rodrigoi** [comment]
@dmsnell answering the last two questions, this has to live side by side with the current implementation under `comments` until all components are wired into the store via `connect`, and cleared of `Immutable` instances, that still a big player in the comments components thanks to `buildCommentsTree` on `client/state/comments/utils`.

All this should take over `comments` one component at a time, starting with the action buttons, and in the end we'll be able to rename it back to `comments`.

@gwwar this is ready for another review, with new tests and the data layer abstraction for these actions.

File: client/state/data-layer/wpcom/discussions/content/index.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  * @returns {Promise}           Promise
+  */
+export const requestCommentContentUpdate = ( { dispatch }, action, next ) => {
+	const {
+		siteId,
+		commentId,
+		content
+	} = action;
+
+	log( 'Request content update for comment %d on site %d using content %o', commentId, siteId, content );
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/discussions/content/index.js]: please note that we will have a record of this in the Redux action log so this and similar log statements may not be necessary. 😄

**Comment by dmsnell** [comment]
@rodrigoi do we think we could eliminate Immutable first in a series of PRs which update the selectors/reducers? I'm nervous about duplicating the comments in the state tree, renaming in the future, and introducing confusion into the code in the transition time.

> has to live side by side

^^^ why does it have to? is there some reason we can't incrementally remove Immutable?

File: client/state/action-types.js [hunk 2]
```diff
@@ -89,8 +89,26 @@ export const DIRECTLY_ASK_QUESTION = 'DIRECTLY_ASK_QUESTION';
...
+export const DISCUSSIONS_ITEM_STATUS_UPDATE_REQUESTING = 'DISCUSSIONS_ITEM_STATUS_UPDATE_REQUESTING';
+export const DISCUSSIONS_ITEM_STATUS_UPDATE_REQUEST_FAILURE = 'DISCUSSIONS_ITEM_STATUS_UPDATE_REQUEST_FAILURE';
+export const DISCUSSIONS_ITEM_STATUS_UPDATE_REQUEST_SUCCESS = 'DISCUSSIONS_ITEM_STATUS_UPDATE_REQUEST_SUCCESS';
+export const DISCUSSIONS_RECEIVE = 'DISCUSSIONS_RECEIVE';
+export const DISCUSSIONS_REQUEST = 'DISCUSSIONS_REQUEST';
+export const DISCUSSIONS_REQUEST_FAILURE = 'DISCUSSIONS_REQUEST_FAILURE';
+export const DISCUSSIONS_REQUEST_SUCCESS = 'DISCUSSIONS_REQUEST_SUCCESS';
+export const DISCUSSIONS_REQUESTING = 'DISCUSSIONS_REQUESTING';
+export const DOCUMENT_HEAD_LINK_ADD = 'DOCUMENT_HEAD_LINK_ADD';
+export const DOCUMENT_HEAD_META_ADD = 'DOCUMENT_HEAD_META_ADD';
```
Comment (gwwar) [inline client/state/action-types.js]: Did we need `DOCUMENT_HEAD_LINK_ADD` and `DOCUMENT_HEAD_LINK_ADD`?

File: client/state/data-layer/wpcom/discussions/index.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		siteId,
+		postId,
+		status
+	} = action;
+
+	log( 'Request discussions for post %d on site %d using status %o', postId, siteId, status );
+
+	dispatch( requestingPostComments( siteId, postId, status ) );
+
+	return wpcom.site( siteId )
```
Comment (gwwar) [inline client/state/data-layer/wpcom/discussions/index.js]: Looks like these actions aren't using the http layer. Would it be easier if we scoped down the initial PR to the actions you already converted? It'll be easier to test and faster for this to land.

**Comment by gwwar** [comment]
Closing this one out.

**Pull request closed without merging**